src/language_features/hover.rs

```diff
@@ -105,43 +116,112 @@ pub fn editor_hover(
+                MarkupKind::Markdown => {
+                    if let HoverType::InfoInHoverClient { .. } = hover_type {
+                        contents.value
+                    } else {
+                        markdown_to_kakoune_markup(contents.value, force_plaintext)
+                    }
+                }
                 MarkupKind::PlainText => contents.value,
             },
         },
     };
 
-    if contents.is_empty() && diagnostics.is_empty() && maybe_hover_modal.is_none() {
-        return;
-    }
+    match hover_type {
+        HoverType::Normal => {
+            if contents.is_empty() && diagnostics.is_empty() {
+                return;
+            }
 
-    let anchor_or_style = if let Some(hover_modal) = maybe_hover_modal.as_ref() {
-        contents = format!("{}\n---\n{}", hover_modal.context, contents);
-        "modal".to_string()
-    } else {
-        format!("{}", params.position)
+            let command = format!(
+                "lsp-show-hover {} %§{}§ %§{}§",
+                params.position,
+                contents.replace("§", "§§"),
+                diagnostics.replace("§", "§§"),
+            );
+            ctx.exec(meta, command);
+        }
+        HoverType::Modal {
+            modal_heading,
+            do_after,
+        } => {
+            show_hover_modal(meta, ctx, modal_heading, do_after, contents, diagnostics);
+        }
+        HoverType::InfoInHoverClient { fifo } => {
+            show_hover_in_hover_client(meta, ctx, fifo, contents);
+        }
     };
+}
 
-    let mut command = format!(
-        "lsp-show-hover {} %§{}§ %§{}§",
-        anchor_or_style,
+fn show_hover_modal(
+    meta: EditorMeta,
+    ctx: &Context,
+    modal_heading: String,
+    do_after: String,
+    contents: String,
+    diagnostics: String,
+) {
+    let contents = format!("{}\n---\n{}", modal_heading, contents);
+    let command = format!(
+        "lsp-show-hover modal %§{}§ %§{}§",
         contents.replace("§", "§§"),
-        diagnostics.replace("§", "§§")
+        diagnostics.replace("§", "§§"),
     );
-
-    // Wrap if we're using a HoverModal
-    if let Some(hover_modal) = maybe_hover_modal {
-        command = formatdoc!(
-            "eval %§
+    let wrapped_command = formatdoc!(
+        "eval %§
                  {}
                  {}
              §",
-            command.replace("§", "§§"),
-            hover_modal.do_after.replace("§", "§§")
-        );
+        command.replace("§", "§§"),
+        do_after.replace("§", "§§")
+    );
+    ctx.exec(meta, wrapped_command);
+}
+
+fn show_hover_in_hover_client(
+    meta: EditorMeta,
+    ctx: &Context,
+    hover_fifo: String,
+    contents: String,
+) {
+    if contents.is_empty() {
+        return;
     }
 
+    let handle = std::thread::spawn(move || {
+        // Note that we don't show diagnostics in the hover client
+        fs::write(hover_fifo, contents.as_bytes()).unwrap();
+    });
+
+    let command = formatdoc!(
```

Comment:
if there are no arguments to format, I guess indoc! is preferred over formatdoc!

*However*, since this command is used twice with different indentation, I'd just write it in one line, to avoid weird-looking logs
Something like

```rust
    let command = "%[ \
            edit! -existing -readonly -fifo %opt[lsp_hover_fifo] *hover*; \
            set buffer=*hover* filetype markdown; \
            ]";
```